File: leetcode/array/16.py
from typing import List


class Solution:

    def threeSumClosest(self, nums: List[int], target: int) -> int:
        '''最接近的三数和
        
        @Note:
            排序->三指针,固定一个,双边遍历. 添加一个绝对差值判断
        '''
        lens = len(nums)
        nums = sorted(nums)
        diff = 10000
        ret = None
        for i in range(lens):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            l = i + 1
            r = lens - 1
            while l < r:
                sum = nums[i] + nums[l] + nums[r]
                if abs(sum - target) < diff:
                    diff = abs(sum - target)
                    ret = sum
                if sum == target:
                    return sum
                elif sum > target:
                    r -= 1
                elif sum < target:
                    l += 1
        return ret


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    target = int(input().strip())
    solution = Solution()
    print(solution.threeSumClosest(arr, target))

File: leetcode/array/16.cpp
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
 public:
  int three_sum_closest(std::vector<int>& arr, int target) {
    int ret = -1;
    std::sort(arr.begin(), arr.end());

    float diff = std::numeric_limits<float>::max();
    for (size_t i = 0; i < arr.size(); ++i) {
      if (i > 0 && arr[i] == arr[i - 1]) {  // 去重
        continue;
      }
      size_t l = i + 1, r = arr.size() - 1;
      while (l < r) {
        int sum = arr[i] + arr[l] + arr[r];
        if (std::abs(sum - target) < diff) {
          diff = std::abs(sum - target);
          ret = sum;
        }
        if (sum > 0) {
          --r;
        } else if (sum < 0) {
          ++l;
        } else {
          return sum;
        }
      }
    }
    return ret;
  }
};

int main() {
  std::vector<int> a = {1, 5, 0, 2, 2, 3};

  Solution solution = Solution();
  std::vector<std::vector<int>> ret;
  std::cout << solution.three_sum_closest(a, 4);
}